### PR TITLE
New: Support for Arrow Functions

### PR DESCRIPTION
**Pull request by xjamundx**
Support for arrow functions including default params.

Most of this was ported from https://github.com/jquery/esprima/commit/c6bb861b40ffcc90a6f21931d77ecaea10c22f05

- [x] added all of the latest tests from esprima 2.0
- [x] needs to error when `ecmaFeatures.arrowFunctions` is false
- [x] tests need to pass (all are failing right now)
- [x] need to also port this over (default params) https://github.com/jquery/esprima/commit/e49e897c7c05eb9bda34b524bc85caf6ece6274b
- [x] add the tests from the errors section https://github.com/jquery/esprima/commit/c6bb861b40ffcc90a6f21931d77ecaea10c22f05#diff-c1129c8b045390789fa8ff62f2c6b4a9R19724

### CODE REVIEW

File: lib/features.js [hunk 1]
```diff
@@ -40,6 +40,9 @@
 
 module.exports = {
 
+	// enable parsing of arrow functions
```
Comment (nzakas) [inline lib/features.js]: Should be true to turn on.

Also, is this indented too far?

**Comment by nzakas** [comment]
Looks like a good start!

File: lib/features.js (hunk 1 above)
Comment (xjamundx) [inline lib/features.js]: yeah will fix spacing, there are a few other spacing issues (still)

**Comment by xjamundx** [comment]
Hoping to be able to finish this up in the few days!

**Comment by nzakas** [comment]
Awesome! Throngs of ESLint fans will love you.

**Comment by xjamundx** [comment]
Chasing down 2 types of errors now:

This is the I'm probably doing something wrong with `markerApply()` thing.

```
TypeError: Cannot read property 'offset' of undefined
    at markerApply (/Users/jamuferguson/dev/espree/espree.js:9:89606)
    at parseArrowFunctionExpression (/Users/jamuferguson/dev/espree/espree.js:9:148717)
    at parseAssignmentExpression (/Users/jamuferguson/dev/espree/espree.js:9:150278)
    at parseExpression (/Users/jamuferguson/dev/espree/espree.js:9:153872)
    at parseExpressionStatement (/Users/jamuferguson/dev/espree/espree.js:9:159543)
    at parseStatement (/Users/jamuferguson/dev/espree/espree.js:9:184409)
```

And then this is the I'm totally missing some parsing somewhere issue:

```
Error: Line 1: Unexpected token =>
    at throwError (/Users/jamuferguson/dev/espree/espree.js:9:97289)
    at throwUnexpected (/Users/jamuferguson/dev/espree/espree.js:9:100904)
    at expect (/Users/jamuferguson/dev/espree/espree.js:9:101346)
    at parseGroupExpression (/Users/jamuferguson/dev/espree/espree.js:9:122308)
    at parsePrimaryExpression (/Users/jamuferguson/dev/espree/espree.js:9:122694)
    at parseLeftHandSideExpressionAllowCall (/Users/jamuferguson/dev/espree/espree.js:9:130115)
    at parsePostfixExpression (/Users/jamuferguson/dev/espree/espree.js:9:133260)
    at parseUnaryExpression (/Users/jamuferguson/dev/espree/espree.js:9:138431)
    at parseBinaryExpression (/Users/jamuferguson/dev/espree/espree.js:9:141985)
    at parseConditionalExpression (/Users/jamuferguson/dev/espree/espree.js:9:144783)
    at parseAssignmentExpression (/Users/jamuferguson/dev/espree/espree.js:9:150522)
    at parseExpression (/Users/jamuferguson/dev/espree/espree.js:9:153872)
    at parseExpressionStatement (/Users/jamuferguson/dev/espree/espree.js:9:159543)
    at parseStatement (/Users/jamuferguson/dev/espree/espree.js:9:184409)
    at parseSourceElement (/Users/jamuferguson/dev/espree/espree.js:9:207847)
    at parseSourceElements (/Users/jamuferguson/dev/espree/espree.js:9:209833)
    at parseProgram (/Users/jamuferguson/dev/espree/espree.js:9:210602)
```

**Comment by nzakas** [comment]
Cool, I was going to suggest looking at current Esprima harmony branch, as some of the functionality may have changed since the first commit.

**Comment by xjamundx** [comment]
Main tests pass, now stuck on:
```
runner: 663 tests. 1 failures. 436 ms
Syntax: Expected
    {
        "AssignmentExpression": "AssignmentExpression",
        "ArrayExpression": "ArrayExpression",
        "BlockStatement": "BlockStatement",
        "BinaryExpression": "BinaryExpression",
        "BreakStatement": "BreakStatement",
        ....
Diff:
 {
+  ArrowFunctionExpression: "ArrowFunctionExpression"
 }
```

**Comment by nzakas** [comment]
This: https://github.com/eslint/espree/blob/master/tests/fixtures/ast/API.json

**Comment by xjamundx** [comment]
:+1:

**Comment by xjamundx** [comment]
Back to the real world for a bit, will try to add default-params (using the mix thing) support today, rebase it down and then. yep. then party!

File: lib/features.js [hunk 2]
```diff
@@ -40,6 +40,9 @@
 
 module.exports = {
 
+    // enable parsing of arrow functions
+    arrowFunctions: false,
```
Comment (xjamundx) [inline lib/features.js]: Was this supposed to be `true`? Most are false

File: espree.js [hunk 3]
```diff
@@ -58,6 +59,13 @@ var Token = tokenInfo.Token,
     state,
     extra;
 
+// TODO: move this to its own file or something
```
Comment (xjamundx) [inline espree.js]: I think I'll just remove this comment

File: lib/features.js (hunk 2 above)
Comment (nzakas) [inline lib/features.js]: It's true in the readme, false in features.js

**Comment by nzakas** [comment]
Looks good. Go ahead and remove the comment.

**Comment by nzakas** [comment]
Sweet, thanks!

**Pull request merged**